cpp/add_synonyms.cc

```diff
@@ -0,0 +1,204 @@
+
+    MarcXmlWriter xml_writer(marc_output);
+
+    while (MarcUtil::Record record = MarcUtil::Record::XmlFactory(marc_input)) {
+        ProcessRecord(&record, synonym_map, primary_tags_and_subfield_codes, output_tags_and_subfield_codes);
+        record.write(&xml_writer);
+        ++record_count;
+    }
+
+    std::cerr << "Modified " << modified_count << " of " << record_count << " record(s).\n";
+}
+
+int main(int argc, char **argv) {
+    progname = argv[0];
+
+    if (argc != 4)
+        Usage();
+
+    const std::string marc_input_filename(argv[1]);
+    std::unique_ptr<File> marc_input(FileUtil::OpenInputFileOrDie(marc_input_filename));
+
+    const std::string norm_data_marc_input_filename(argv[2]);
+    std::unique_ptr<File> norm_data_marc_input(FileUtil::OpenInputFileOrDie(norm_data_marc_input_filename));
+
+    const std::string marc_output_filename(argv[3]);
+    if (unlikely(marc_input_filename == marc_output_filename))
+        Error("Title data input file name equals output file name!");
+    if (unlikely(norm_data_marc_input_filename == marc_output_filename))
+        Error("Authority data input file name equals output file name!");
+
+    std::string output_mode("w");
+    if (marc_input->isCompressingOrUncompressing())
+        output_mode += 'c';
+    File marc_output(marc_output_filename, output_mode);
+    if (not marc_output)
+        Error("can't open \"" + marc_output_filename + "\" for writing!");
+
+    try {
+        // Determine possible mappings
+        const std::string AUTHORITY_DATA_PRIMARY_SPEC = "110abcd:111abcd:130abcd:150abcd:151abcd";
+        const std::string AUTHORITY_DATA_SYNONYM_SPEC = "410abcd:411abcd:430abcd:450abcd:451abcd";
+        const std::string TITLE_DATA_PRIMARY_SPEC = "610abcd:611abcd:430abcd:650abcd:651abcd";
+        const std::string TITLE_DATA_UNUSED_FIELDS_FOR_SYNONYMS = "180a:181a:182a:183a:184a";
+
+        // Determine fields to handle
+        std::set<std::string> primary_tags_and_subfield_codes;
+        std::set<std::string> synonym_tags_and_subfield_codes;
+        std::set<std::string> input_tags_and_subfield_codes;
+        std::set<std::string> output_tags_and_subfield_codes;
+
+        if (unlikely(StringUtil::Split(AUTHORITY_DATA_PRIMARY_SPEC, ":", &primary_tags_and_subfield_codes) < 1))
+            Error("Need at least one primary field");
+
+        if (unlikely(StringUtil::Split(AUTHORITY_DATA_SYNONYM_SPEC, ":", &synonym_tags_and_subfield_codes) < 1))
+            Error("Need at least one synonym field");
+
+        if (unlikely(StringUtil::Split(TITLE_DATA_PRIMARY_SPEC, ":", &input_tags_and_subfield_codes) < 1))
+            Error("Need at least one input field");
+
+        if (unlikely(StringUtil::Split(TITLE_DATA_UNUSED_FIELDS_FOR_SYNONYMS, ":", &output_tags_and_subfield_codes) < 1))
+            Error("Need at least one output field");
```

Comment:
If splitting on a single character it is more efficient to use a character instead of a string.